File: Mentoring/랜덤코딩디펜스/TarjanAlgorithm.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline
id = 0
def dfs_cost(cur, visited):
    global tmp
    for nxt in graph[cur]:
        if nxt in scc and not visited[nxt]:
            visited[nxt] = True
            tmp += cost[cur][nxt]
            dfs_cost(nxt, visited)

def dfs(cur):
    global id
    id += 1
    visited[cur] = id
    stack.append(cur)
    on_stack[cur] = True
    parent = visited[cur]
    for nxt in graph[cur]:
        if visited[nxt] == -1:
            parent = min(parent, dfs(nxt))
        elif on_stack[nxt]:
            parent = min(parent, visited[nxt])
    
    if parent == visited[cur]:
        scc = set()
        while True:
            x = stack.pop()
            on_stack[x] = False
            scc.add(x)
            if x == cur:
                break

        result.append(scc)

    return parent
T = int(input())
for test_case in range(1, T+1):
    V, E = map(int, input().split())
    visited = [-1]*(V+1)
    on_stack = [False]*(V+1)
    stack = []
    graph = [[] for _ in range(V+1)]
    cost = [[0]*(V+1) for _ in range(V+1)]
    for _ in range(E):
        a, b, x = map(int, input().split())
        graph[a].append(b)
        cost[a][b] = x

    result = []
    for i in range(1, V+1):
        if visited[i] == -1:
            dfs(i)

    res = sys.maxsize
    for scc in result:
        tmp = 0
        visited = [False]*(V+1)
        dfs_cost(list(scc)[0], visited)
        if tmp:
            res = min(res, tmp)

    print(f'#{test_case}', res)

File: Mentoring/랜덤코딩디펜스/TarjanAlgorithm.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;
import java.util.StringTokenizer;

public class TarjanAlgorithm {
    static int id = 0;
    static List<Integer>[] graph;
    static int[] visited;
    static boolean[] onStack;
    static Stack<Integer> stack;
    static List<List<Integer>> result;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int V = Integer.parseInt(st.nextToken());
        int E = Integer.parseInt(st.nextToken());
        visited = new int[V + 1];
        onStack = new boolean[V + 1];
        stack = new Stack<>();
        graph = new ArrayList[V + 1];
        for (int i = 1; i <= V; i++) {
            graph[i] = new ArrayList<>();
        }
        result = new ArrayList<>();

        for (int i = 0; i < E; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            graph[a].add(b);
        }

        for (int i = 1; i <= V; i++) {
            if (visited[i] == 0) {
                dfs(i);
            }
        }

        System.out.println(result.size());
        result.sort((a, b) -> a.get(0).compareTo(b.get(0)));
        for (List<Integer> list : result) {
            for (int num : list) {
                System.out.print(num + " ");
            }
            System.out.println();
        }
    }

    private static int dfs(int cur) {
        id++;
        visited[cur] = id;
        stack.push(cur);
        onStack[cur] = true;
        int parent = visited[cur];

        for (int nxt : graph[cur]) {
            if (visited[nxt] == 0) {
                parent = Math.min(parent, dfs(nxt));
            } else if (onStack[nxt]) {
                parent = Math.min(parent, visited[nxt]);
            }
        }

        if (parent == visited[cur]) {
            List<Integer> css = new ArrayList<>();
            while (true) {
                int x = stack.pop();
                onStack[x] = false;
                css.add(x);
                if (x == cur) {
                    break;
                }
            }
            css.sort(null);
            css.add(-1);
            result.add(css);
        }

        return parent;
    }
}
